File: app/src/coupon/smart.js
// AKILLI KUPON + ANALİZDEN AKTARIM — saf modül (RN/api importu yok; testli).
// DÜRÜSTLÜK KURALLARI (kesin):
// • Veri yoksa uydurulmaz: sinyali olmayan maç "veri yetersiz" işaretlenir ve
//   boş bırakılır — rastgele/uydurma seçim üretilmez.
// • Kapsama puanı KESİN KAZANMA İHTİMALİ DEĞİLDİR — her yerde bu notla sunulur.
// • Aktarım kullanıcı seçimini ASLA sessizce değiştirmez: önce fark listesi
//   (diffSelections) gösterilir, karar kullanıcınındır.
// • Kolon sınırı 2500 (resmi kural) hiçbir durumda aşılmaz.
import { COUPON_MAX_COLUMNS } from '../couponConfig';

export const expandSymbol = (sym) =>
  (sym && sym !== '-' ? String(sym).split('').map((c) => (c === '0' ? 'X' : c)).filter((c) => ['1', 'X', '2'].includes(c)) : []);

// Maçtan MEVCUT sinyalleri topla — olmayan alan null kalır, uydurulmaz.
export function signalsOf(m) {
  const probs = m?.probabilities && ['1', 'X', '2'].every((k) => Number.isFinite(Number(m.probabilities[k])))
    ? { 1: Number(m.probabilities['1']), X: Number(m.probabilities['X']), 2: Number(m.probabilities['2']) }
    : null;
  const sysSym = m?.prediction?.symbol && m.prediction.symbol !== '-' ? m.prediction.symbol : null;
  const master = m?.radarCenter?.master || null;
  const radarSym = master?.favorite?.symbol || null;
  const surprise = Number.isFinite(Number(m?.analysis?.surpriseScore)) ? Number(m.analysis.surpriseScore) : null;
  const dataQuality = Number.isFinite(Number(master?.dataQuality)) ? Number(master.dataQuality) : null;
  return { probs, sysSym, radarSym, surprise, dataQuality };
}

// İhtimalleri büyükten küçüğe sırala → [['1',45],['X',30],['2',25]]
function rankedProbs(probs) {
  return Object.entries(probs).sort((a, b) => b[1] - a[1]);
}

// Belirsizlik (0-1) + insanca gerekçeler. Yalnız MEVCUT sinyaller katılır.
export function uncertaintyOf(sig) {
  const parts = []; const reasons = [];
  if (sig.probs) {
    const r = rankedProbs(sig.probs);
    const gap = (r[0][1] - r[1][1]) / 100;                 // favori ile ikinci arasındaki fark
    const u = Math.max(0, Math.min(1, 1 - gap * 2));       // fark küçükse belirsiz
    parts.push({ w: 3, v: u });
    if (gap <= 0.10) reasons.push('ihtimaller birbirine çok yakın');
    else if (gap >= 0.25) reasons.push('güçlü bir aday var');
  }
  if (sig.surprise != null) {
    parts.push({ w: 2, v: Math.max(0, Math.min(1, sig.surprise / 100)) });
    if (sig.surprise >= 65) reasons.push('sürpriz riski yüksek');
  }
  const sysFirst = expandSymbol(sig.sysSym)[0] || null;
  if (sysFirst && sig.radarSym) {
    const disagree = sysFirst !== sig.radarSym;
    parts.push({ w: 2, v: disagree ? 1 : 0 });
    if (disagree) reasons.push('Master Analiz ile Radar farklı yönde');
    else reasons.push('Master Analiz ile Radar aynı yönde');
  }
  if (sig.dataQuality != null) {
    const v = Math.max(0, Math.min(1, 1 - sig.dataQuality / 100));
    parts.push({ w: 1, v });
    if (sig.dataQuality < 50) reasons.push('veri yeterliliği düşük');
  }
  if (!parts.length) return { u: null, reasons: ['bu maç için sinyal verisi yok'] };
  const u = parts.reduce((t, p) => t + p.w * p.v, 0) / parts.reduce((t, p) => t + p.w, 0);
  return { u, reasons };
}

// Maçın taban (tekli) tercihi + kaynağı. Veri yoksa null (uydurulmaz).
export function baseOutcomeOf(sig) {
  if (sig.probs) return { outcome: rankedProbs(sig.probs)[0][0], source: 'ihtimaller' };
  if (sig.radarSym) return { outcome: sig.radarSym, source: 'Radar favorisi' };
  const sys = expandSymbol(sig.sysSym);
  if (sys.length) return { outcome: sys[0], source: 'Sistem tahmini' };
  return null;
}
// Bir sonraki en olası işaret (genişletme adayı).
function nextOutcomeOf(sig, chosen) {
  if (sig.probs) {
    for (const [o] of rankedProbs(sig.probs)) if (!chosen.includes(o)) return o;
  }
  const sys = expandSymbol(sig.sysSym);
  for (const o of sys) if (!chosen.includes(o)) return o;
  if (sig.radarSym && !chosen.includes(sig.radarSym)) return sig.radarSym;
  return ['1', 'X', '2'].find((o) => !chosen.includes(o)) || null;
}

// AKILLI KUPON: bütçe (kolon cinsinden) + hedef (12-15) → seçim + açıklamalar.
// Mantık: tüm maçlar tabanla tekli başlar; en belirsiz maçlar, bütçe ve 2500
// sınırı içinde kalarak çifte/üçlüye genişletilir. Hedef düştükçe (12'ye doğru)
// genişletme iştahı azalır (daha çok yanlışa tahammül var → daha ucuz kupon).
export function buildSmartCoupon({ matches, budgetColumns, target }) {
  const maxCols = Math.max(1, Math.min(COUPON_MAX_COLUMNS, Math.floor(budgetColumns || COUPON_MAX_COLUMNS)));
  const tgt = [12, 13, 14, 15].includes(target) ? target : 13;
  // Hedefe göre genişletme eşiği: 15 → her belirsizlik değerli; 12 → yalnız çok belirsizler.
  const threshold = { 15: 0.15, 14: 0.30, 13: 0.45, 12: 0.60 }[tgt];

  const rows = (matches || []).map((m) => {
    const sig = signalsOf(m);
    const unc = uncertaintyOf(sig);
    const base = baseOutcomeOf(sig);
    return { no: m.no, sig, u: unc.u, reasons: unc.reasons, base };
  });
  const insufficient = rows.filter((r) => !r.base).map((r) => r.no);
  const usable = rows.filter((r) => r.base);

  const picks = new Map(usable.map((r) => [r.no, [r.base.outcome]]));
  const colsOf = () => [...picks.values()].reduce((n, o) => n * o.length, 1);

  // Açgözlü genişletme: en belirsiz + eşiği aşan maçtan başla.
  const queue = usable.filter((r) => r.u != null && r.u >= threshold).sort((a, b) => b.u - a.u);
  let guard = 0;
  while (guard++ < 60) {
    let applied = false;
    for (const r of queue) {
      const cur = picks.get(r.no);
      if (cur.length >= 3) continue;
      // Üçlüye yalnız ÇOK belirsiz maçta çık (u ≥ 0.75) — savurganlık değil sinyal.
      if (cur.length === 2 && r.u < 0.75) continue;
      const nxt = nextOutcomeOf(r.sig, cur);
      if (!nxt) continue;
      const factor = (cur.length + 1) / cur.length;
      if (colsOf() * factor > maxCols) continue;
      picks.set(r.no, [...cur, nxt]);
      applied = true;
      break; // her turda tek genişletme → en değerliye öncelik korunur
    }
    if (!applied) break;
  }

  const ORDER = ['1', 'X', '2'];
  const selections = usable.map((r) => ({
    no: r.no,
    selectedOutcomes: ORDER.filter((o) => picks.get(r.no).includes(o)),
  }));

  // Sade açıklamalar — teknik formül yok, insanca gerekçe var.
  const explanations = usable.map((r) => {
    const n = picks.get(r.no).length;
    const level = n === 1 ? 'tekli' : n === 2 ? 'çifte' : 'üçlü';
    const why = r.reasons.slice(0, 2).join(' · ') || 'sinyaller tek yönde';
    const head = n === 1
      ? (r.u != null && r.u >= threshold ? 'tekli kaldı (bütçe/sınır önceliği başka maçlardaydı)' : 'tekli bırakıldı')
      : `${level} yapıldı`;
    return { no: r.no, level, text: `${head} — ${why} (kaynak: ${r.base.source})` };
  });

  // Kapsama puanı: seçilen işaretlerin ihtimal kütlesi ortalaması (yalnız
  // ihtimal verisi olan maçlarda). KESİN KAZANMA İHTİMALİ DEĞİLDİR.
  const covered = usable.filter((r) => r.sig.probs);
  const coverageScore = covered.length
    ? Math.round(covered.reduce((t, r) => t + picks.get(r.no).reduce((s, o) => s + r.sig.probs[o], 0), 0) / covered.length)
    : null;

  return {
    selections,
    columns: colsOf(),
    target: tgt,
    explanations,
    coverageScore,
    coverageNote: 'Kapsama puanı seçimlerin sinyal ihtimallerini ne kadar örttüğünü gösterir — kesin kazanma ihtimali DEĞİLDİR.',
    insufficient, // veri yetersiz maçlar — boş bırakıldı, kullanıcı elle seçer
  };
}

// AKTARIM FARKI — mevcut seçim ASLA sessizce ezilmez; fark listesi kullanıcıya
// gösterilir, onaylarsa uygulanır.
export function diffSelections(currentMap, proposedMap) {
  const nos = new Set([...Object.keys(currentMap || {}), ...Object.keys(proposedMap || {})].map(Number));
  const changes = [];
  for (const no of [...nos].sort((a, b) => a - b)) {
    const from = (currentMap?.[no] || []).join('-');
    const to = (proposedMap?.[no] || []).join('-');
    if (from === to) continue;
    changes.push({
      no,
      from: from || '(boş)',
      to: to || '(boş)',
      kind: !from ? 'fill' : 'change', // fill: boş dolduruldu · change: mevcut değişecek
    });
  }
  return changes;
}

// Aktarım önerisi: Sistem tahmini veya Radar (favori + yakın ikinciyle çifte).
export function proposalFrom(matches, source /* 'system' | 'radar' */) {
  const out = {};
  for (const m of matches || []) {
    const sig = signalsOf(m);
    if (source === 'system') {
      const o = expandSymbol(sig.sysSym);
      if (o.length) out[m.no] = o;
    } else if (source === 'radar') {
      if (!sig.radarSym) continue;
      const arr = [sig.radarSym];
      // Alternatif: ihtimal verisi varsa ve ikinci işaret favoriye ≤10 puan yakınsa çifte öner.
      if (sig.probs) {
        const r = rankedProbs(sig.probs);
        const second = r.find(([o]) => o !== sig.radarSym);
        if (second && sig.probs[sig.radarSym] != null && second[1] >= sig.probs[sig.radarSym] - 10 && second[1] >= 25) arr.push(second[0]);
      }
      out[m.no] = ['1', 'X', '2'].filter((o) => arr.includes(o));
    }
  }
  return out;
}

File: flutter/lib/core/coupon/smart.dart
// KAYNAK: app/src/coupon/smart.js — BİREBİR çeviri.
//
// AKILLI KUPON + ANALİZDEN AKTARIM — saf modül (çizim/api importu yok; testli).
//
// DÜRÜSTLÜK KURALLARI (kesin):
// • Veri yoksa uydurulmaz: sinyali olmayan maç "veri yetersiz" işaretlenir ve
//   boş bırakılır — rastgele/uydurma seçim üretilmez.
// • Kapsama puanı KESİN KAZANMA İHTİMALİ DEĞİLDİR — her yerde bu notla sunulur.
// • Aktarım kullanıcı seçimini ASLA sessizce değiştirmez: önce fark listesi
//   (diffSelections) gösterilir, karar kullanıcınındır.
// • Kolon sınırı 2500 (resmi kural) hiçbir durumda aşılmaz.

import 'dart:math' as math;

import 'coupon_config.dart';

const List<String> _kOrder = ['1', 'X', '2'];

/// Kupon sembolünü (1/0/2/10/02/12/102) olası sonuç listesine açar.
/// 0 = X (beraberlik). '-' veya boş → boş liste.
List<String> expandSymbol(Object? sym) {
  if (sym == null) return const [];
  final s = '$sym';
  if (s.isEmpty || s == '-') return const [];
  return [
    for (final c in s.split(''))
      if (_kOrder.contains(c == '0' ? 'X' : c)) (c == '0' ? 'X' : c),
  ];
}

double? _sonlu(Object? v) {
  if (v is num) return v.isFinite ? v.toDouble() : null;
  final n = double.tryParse('$v');
  return (n != null && n.isFinite) ? n : null;
}

typedef Signals = ({
  Map<String, double>? probs,
  String? sysSym,
  String? radarSym,
  double? surprise,
  double? dataQuality,
});

/// Maçtan MEVCUT sinyalleri topla — olmayan alan null kalır, uydurulmaz.
Signals signalsOf(Map? m) {
  final p = m?['probabilities'] as Map?;
  final probs = (p != null && _kOrder.every((k) => _sonlu(p[k]) != null))
      ? {for (final k in _kOrder) k: _sonlu(p[k])!}
      : null;

  final pred = m?['prediction'] as Map?;
  final sysSym = (pred?['symbol'] != null && pred!['symbol'] != '-')
      ? '${pred['symbol']}'
      : null;

  final master = (m?['radarCenter'] as Map?)?['master'] as Map?;
  final radarSym = (master?['favorite'] as Map?)?['symbol'] as String?;

  final surprise = _sonlu((m?['analysis'] as Map?)?['surpriseScore']);
  final dataQuality = _sonlu(master?['dataQuality']);

  return (
    probs: probs,
    sysSym: sysSym,
    radarSym: radarSym,
    surprise: surprise,
    dataQuality: dataQuality,
  );
}

/// İhtimalleri büyükten küçüğe sırala → [('1',45),('X',30),('2',25)]
List<(String, double)> _rankedProbs(Map<String, double> probs) {
  // JS `Object.entries` ekleme sırasını korur; `signalsOf` her zaman 1/X/2
  // sırasıyla kurar, bu yüzden eşit değerlerde sıra 1 → X → 2 olur.
  final e = [for (final k in _kOrder) (k, probs[k]!)];
  // JS `sort` kararlıdır; eşitlikte ilk gelen önde kalsın diye indeksle.
  final indeksli = [for (var i = 0; i < e.length; i++) (i: i, v: e[i])]
    ..sort((a, b) {
      final c = b.v.$2.compareTo(a.v.$2);
      return c != 0 ? c : a.i.compareTo(b.i);
    });
  return [for (final x in indeksli) x.v];
}

typedef Uncertainty = ({double? u, List<String> reasons});

/// Belirsizlik (0-1) + insanca gerekçeler. Yalnız MEVCUT sinyaller katılır.
Uncertainty uncertaintyOf(Signals sig) {
  final parts = <({double w, double v})>[];
  final reasons = <String>[];

  if (sig.probs != null) {
    final r = _rankedProbs(sig.probs!);
    // favori ile ikinci arasındaki fark
    final gap = (r[0].$2 - r[1].$2) / 100;
    // fark küçükse belirsiz
    final u = math.max(0.0, math.min(1.0, 1 - gap * 2));
    parts.add((w: 3, v: u));
    if (gap <= 0.10) {
      reasons.add('ihtimaller birbirine çok yakın');
    } else if (gap >= 0.25) {
      reasons.add('güçlü bir aday var');
    }
  }

  if (sig.surprise != null) {
    parts.add((w: 2, v: math.max(0.0, math.min(1.0, sig.surprise! / 100))));
    if (sig.surprise! >= 65) reasons.add('sürpriz riski yüksek');
  }

  final sys = expandSymbol(sig.sysSym);
  final sysFirst = sys.isNotEmpty ? sys.first : null;
  if (sysFirst != null && sig.radarSym != null) {
    final disagree = sysFirst != sig.radarSym;
    parts.add((w: 2, v: disagree ? 1 : 0));
    reasons.add(
      disagree
          ? 'Master Analiz ile Radar farklı yönde'
          : 'Master Analiz ile Radar aynı yönde',
    );
  }

  if (sig.dataQuality != null) {
    final v = math.max(0.0, math.min(1.0, 1 - sig.dataQuality! / 100));
    parts.add((w: 1, v: v));
    if (sig.dataQuality! < 50) reasons.add('veri yeterliliği düşük');
  }

  if (parts.isEmpty) {
    return (u: null, reasons: ['bu maç için sinyal verisi yok']);
  }
  final u =
      parts.fold<double>(0, (t, p) => t + p.w * p.v) /
      parts.fold<double>(0, (t, p) => t + p.w);
  return (u: u, reasons: reasons);
}

typedef BaseOutcome = ({String outcome, String source});

/// Maçın taban (tekli) tercihi + kaynağı. Veri yoksa null (uydurulmaz).
BaseOutcome? baseOutcomeOf(Signals sig) {
  if (sig.probs != null) {
    return (outcome: _rankedProbs(sig.probs!).first.$1, source: 'ihtimaller');
  }
  if (sig.radarSym != null) {
    return (outcome: sig.radarSym!, source: 'Radar favorisi');
  }
  final sys = expandSymbol(sig.sysSym);
  if (sys.isNotEmpty) return (outcome: sys.first, source: 'Sistem tahmini');
  return null;
}

/// Bir sonraki en olası işaret (genişletme adayı).
String? _nextOutcomeOf(Signals sig, List<String> chosen) {
  if (sig.probs != null) {
    for (final (o, _) in _rankedProbs(sig.probs!)) {
      if (!chosen.contains(o)) return o;
    }
  }
  for (final o in expandSymbol(sig.sysSym)) {
    if (!chosen.contains(o)) return o;
  }
  if (sig.radarSym != null && !chosen.contains(sig.radarSym)) {
    return sig.radarSym;
  }
  for (final o in _kOrder) {
    if (!chosen.contains(o)) return o;
  }
  return null;
}

typedef SmartSelection = ({Object? no, List<String> selectedOutcomes});
typedef SmartExplanation = ({Object? no, String level, String text});

typedef SmartCoupon = ({
  List<SmartSelection> selections,
  int columns,
  int target,
  List<SmartExplanation> explanations,
  int? coverageScore,
  String coverageNote,
  List<Object?> insufficient,
});

/// AKILLI KUPON: bütçe (kolon cinsinden) + hedef (12-15) → seçim + açıklamalar.
///
/// Mantık: tüm maçlar tabanla tekli başlar; en belirsiz maçlar, bütçe ve 2500
/// sınırı içinde kalarak çifte/üçlüye genişletilir. Hedef düştükçe (12'ye
/// doğru) genişletme iştahı azalır (daha çok yanlışa tahammül var → daha ucuz
/// kupon).
SmartCoupon buildSmartCoupon({
  required List? matches,
  Object? budgetColumns,
  Object? target,
}) {
  final butce = _sonlu(budgetColumns) ?? kCouponMaxColumns.toDouble();
  final maxCols = math.max(
    1,
    math.min(kCouponMaxColumns, butce.floor()),
  );
  final t = target is num ? target.toInt() : int.tryParse('$target');
  final tgt = const [12, 13, 14, 15].contains(t) ? t! : 13;
  // Hedefe göre genişletme eşiği: 15 → her belirsizlik değerli;
  // 12 → yalnız çok belirsizler.
  final threshold = const {15: 0.15, 14: 0.30, 13: 0.45, 12: 0.60}[tgt]!;

  final rows = [
    for (final m in (matches ?? const []).cast<Map>())
      () {
        final sig = signalsOf(m);
        final unc = uncertaintyOf(sig);
        return (
          no: m['no'],
          sig: sig,
          u: unc.u,
          reasons: unc.reasons,
          base: baseOutcomeOf(sig),
        );
      }(),
  ];
  final insufficient = [
    for (final r in rows)
      if (r.base == null) r.no,
  ];
  final usable = [
    for (final r in rows)
      if (r.base != null) r,
  ];

  final picks = <Object?, List<String>>{
    for (final r in usable) r.no: [r.base!.outcome],
  };
  int colsOf() => picks.values.fold<int>(1, (n, o) => n * o.length);

  // Açgözlü genişletme: en belirsiz + eşiği aşan maçtan başla.
  final queue = [
    for (final r in usable)
      if (r.u != null && r.u! >= threshold) r,
  ]..sort((a, b) => b.u!.compareTo(a.u!));

  var guard = 0;
  while (guard++ < 60) {
    var applied = false;
    for (final r in queue) {
      final cur = picks[r.no]!;
      if (cur.length >= 3) continue;
      // Üçlüye yalnız ÇOK belirsiz maçta çık (u ≥ 0.75) — savurganlık değil
      // sinyal.
      if (cur.length == 2 && r.u! < 0.75) continue;
      final nxt = _nextOutcomeOf(r.sig, cur);
      if (nxt == null) continue;
      final factor = (cur.length + 1) / cur.length;
      if (colsOf() * factor > maxCols) continue;
      picks[r.no] = [...cur, nxt];
      applied = true;
      break; // her turda tek genişletme → en değerliye öncelik korunur
    }
    if (!applied) break;
  }

  final selections = [
    for (final r in usable)
      (
        no: r.no,
        selectedOutcomes: [
          for (final o in _kOrder)
            if (picks[r.no]!.contains(o)) o,
        ],
      ),
  ];

  // Sade açıklamalar — teknik formül yok, insanca gerekçe var.
  final explanations = [
    for (final r in usable)
      () {
        final n = picks[r.no]!.length;
        final level = n == 1
            ? 'tekli'
            : n == 2
            ? 'çifte'
            : 'üçlü';
        final why = r.reasons.take(2).join(' · ');
        final head = n == 1
            ? (r.u != null && r.u! >= threshold
                  ? 'tekli kaldı (bütçe/sınır önceliği başka maçlardaydı)'
                  : 'tekli bırakıldı')
            : '$level yapıldı';
        return (
          no: r.no,
          level: level,
          text:
              '$head — ${why.isEmpty ? 'sinyaller tek yönde' : why} '
              '(kaynak: ${r.base!.source})',
        );
      }(),
  ];

  // Kapsama puanı: seçilen işaretlerin ihtimal kütlesi ortalaması (yalnız
  // ihtimal verisi olan maçlarda). KESİN KAZANMA İHTİMALİ DEĞİLDİR.
  final covered = [
    for (final r in usable)
      if (r.sig.probs != null) r,
  ];
  final coverageScore = covered.isEmpty
      ? null
      : (covered.fold<double>(
              0,
              (t, r) =>
                  t +
                  picks[r.no]!.fold<double>(
                    0,
                    (s, o) => s + r.sig.probs![o]!,
                  ),
            ) /
            covered.length)
            .round();

  return (
    selections: selections,
    columns: colsOf(),
    target: tgt,
    explanations: explanations,
    coverageScore: coverageScore,
    coverageNote:
        'Kapsama puanı seçimlerin sinyal ihtimallerini ne kadar örttüğünü '
        'gösterir — kesin kazanma ihtimali DEĞİLDİR.',
    // veri yetersiz maçlar — boş bırakıldı, kullanıcı elle seçer
    insufficient: insufficient,
  );
}

typedef SelectionDiff = ({int no, String from, String to, String kind});

/// AKTARIM FARKI — mevcut seçim ASLA sessizce ezilmez; fark listesi
/// kullanıcıya gösterilir, onaylarsa uygulanır.
List<SelectionDiff> diffSelections(Map? currentMap, Map? proposedMap) {
  final nos = <int>{
    for (final k in (currentMap ?? const {}).keys) ?int.tryParse('$k'),
    for (final k in (proposedMap ?? const {}).keys) ?int.tryParse('$k'),
  }.toList()..sort();

  final changes = <SelectionDiff>[];
  for (final no in nos) {
    final from = _birlestir(currentMap, no);
    final to = _birlestir(proposedMap, no);
    if (from == to) continue;
    changes.add((
      no: no,
      from: from.isEmpty ? '(boş)' : from,
      to: to.isEmpty ? '(boş)' : to,
      // fill: boş dolduruldu · change: mevcut değişecek
      kind: from.isEmpty ? 'fill' : 'change',
    ));
  }
  return changes;
}

/// JS'te anahtar dizgeye çevrilir (`obj[no]`), Dart'ta hem sayı hem dizge
/// anahtar gelebilir — ikisi de aranır.
String _birlestir(Map? map, int no) {
  final v = map?[no] ?? map?['$no'];
  return (v is List) ? v.join('-') : '';
}

/// Aktarım önerisi: Sistem tahmini veya Radar (favori + yakın ikinciyle çifte).
Map<Object?, List<String>> proposalFrom(List? matches, String source) {
  final out = <Object?, List<String>>{};
  for (final m in (matches ?? const []).cast<Map>()) {
    final sig = signalsOf(m);
    if (source == 'system') {
      final o = expandSymbol(sig.sysSym);
      if (o.isNotEmpty) out[m['no']] = o;
    } else if (source == 'radar') {
      if (sig.radarSym == null) continue;
      final arr = <String>[sig.radarSym!];
      // Alternatif: ihtimal verisi varsa ve ikinci işaret favoriye ≤10 puan
      // yakınsa çifte öner.
      if (sig.probs != null) {
        final r = _rankedProbs(sig.probs!);
        final ikinci = r.where((x) => x.$1 != sig.radarSym).firstOrNull;
        final fav = sig.probs![sig.radarSym!];
        if (ikinci != null &&
            fav != null &&
            ikinci.$2 >= fav - 10 &&
            ikinci.$2 >= 25) {
          arr.add(ikinci.$1);
        }
      }
      out[m['no']] = [
        for (final o in _kOrder)
          if (arr.contains(o)) o,
      ];
    }
  }
  return out;
}
